### ISSUE TITLE
Can't add variables to published game with dialogic saves.

### ISSUE THREAD

**Issue by metanoder** [coppolaemilio/dialogic#1278]
### **(Godot 3.5.1 + Dialogic 1.4.5)**

**Problem = Can't add variables to published game with dialogic saves.**

**Description...**
I am prototype testing my visual novel game.  
I am running a test scenario on which I am updating my visual novel after it has been published and out in the wild.  

While using the save feature of dialogic, I am having problems adding new variables to saved games in my early testing.  

**Reproduce...**
1) MacOS
2) Export Project
3) Update game by adding variables, via dialogic IDE **add event** "Set Value", "myNewVar", to itself plus "1"  
4) Export Game.Pck
5) Go to Game DMG in HardDrive, Open DMG and move to a folder.  Right clicking to "Open Package Contents"
6) Game>Contents>Resource>Game.pck 
7) Replace Game.pck 

Say, when a totally new variable called “myNewVar” is introduced after loading a saved game (that did not have “myNewVar” prior), and on continuing, the game progresses that “myNewVar” point, and adds “1” to itself, the game crashes.  

Updating character art and updating timelines seems to work just fine…

But would love to have a way to update / adding new variables to the dialogic .json files and add variable content to deployed games.  This would be also great in the new Godot 4 Dialogic 2 branch.

**Comment by Jowan-Spooner** [comment]
Thanks for the report. This has been an issue for a long time. I currently am very much out of the development of dialogic 1.x. Here are some notes for someone who would like to fix this. Maybe you have time for this @zaknafean?
I think it could be fixed by running a check in `DialogicResources.get_saved_definitions()`:
```gdscript
# return the definition info from the definiiotn.json in the save folder "save name"
static func get_saved_definitions(save_name: String = '') -> Dictionary:
        
        if not save_name in get_saves_folders():
		print("[D] Wasn't able to find save '"+save_name+"'. Loaded the default definitions.")
		return get_default_definitions()
	
        var default_definitions := get_default_definitions()
	
        var saved_definitions := {}
        if save_name == '':
		saved_definitions = load_json(get_config_files_paths()['DEFINITIONS_DEFAULT_SAVE'], default_definitions)
	else:
                saved_definitions = load_json(WORKING_DIR+"/"+save_name+"/definitions.json", {})
        
        for def in default_definitions:
                if not def in saved_definitions:
                       saved_definitions[def] = default_definitions[def]
        return saved_definitions
```

Keep in mind I didn't test this. I just think that would be the right place to put it.

**Comment by zaknafean** [comment]
I'll try to put in a fix for this this weekend

**Comment by metanoder** [comment]
Hi Jowan-Spooner + Zaknafean

Thank you for taking this up. 

TLDR;  Tested the notes you provided (Fantastic btw. Couldn't do without as a novice hacker).  After some of my hacky additions / change, it works with caveats.  I can get duplicates, plus there's a need for deleting depreciated variables on saved games.

My game,( which runs on the dialogic plugin and off of the VN-template)
did start to run smoothly after patching in new variables after inserting your suggested code with tweaks...

- (added) quick check if dictionary keys match, when they don't
- get to the keys within the dictionary nested in an array, nested in a top-level dictionary; used: .variables
- loop, added [] on each dictionary. 
- if it doesn't have, add at index

I ran it over a few saves, and it works.  But I noticed that I can get duplicate variables.  Nor can I erase a variable if I decide to take it out via dialogic editor.  

Still in the dark on how to sort + check the dictionaries.  Plus add a deleting feature for old variables or duplicates.  
But just wanted to share what I've found thus far from your suggestion.  

Many thanks!

`	var default_definitions : Dictionary = get_default_definitions()
	
	var saved_definitions : Dictionary = {}
	
	if save_name == '':
		saved_definitions = load_json(get_config_files_paths()['DEFINITIONS_DEFAULT_SAVE'], default_definitions)
		print("definitions data in resource loaded")
		
	else:
		saved_definitions = load_json(WORKING_DIR+"/"+save_name+"/definitions.json", {})
		print("definitions data on ''user://'' loaded") #gamefiend "this one is getting loaded each time I reload a saved game"
	
	var check = are_dictionary_equal(saved_definitions, default_definitions)
	print(check)
	
	if check == false:
		
		var meta_saved_definitions = saved_definitions.values()
		var meta_default_definitions = default_definitions.values()
		
		
		for i in meta_default_definitions:
			if not i in meta_saved_definitions:
				meta_saved_definitions[0] = meta_default_definitions[0]

	return saved_definitions
	
#my tester from internet
static func are_dictionary_equal(a: Dictionary, b: Dictionary):
	if not a or not b:
		return false
	if a.size() != b.size():
		return false
	for key in a.keys():
		if not b.has(key):
			return false
		if a[key] != b[key]:
			return false

	return true

#not used sorter below.  From internet.  Good for dictionary, but after testing... 
#can't seem to sort dictionary that is nested in an array, nested in another dictionary.

class sorter:
	static func sort_ascending(a, b):
		if a[0] < b[0]:
			return true
		return false`

<img width="1503" alt="get_saved_definitions" src="https://user-images.githubusercontent.com/117393840/200131081-ad9b0c57-9db5-4a06-b1b6-845e302e3a34.png">

**Comment by metanoder** [comment]
Had another go at it.

Probably the loooong way to get around to it, and lots of waste.  I imagine there is a far leaner way to make this happen.
But at the moment, this seems to be working for me.  
changing the variable name is a problem though...

TLDR;
- listed the nested id numbers of each var from saved definitions and default definitions
- sorted them...
- filtered saved definitions id numbers that are not present in default definitions id numbers
- sorted filtered saved definitions id numbers #dunno if I need this...
- added back in default definitions id numbers not in saved definitions id numbers
- filtered saved dictionary, keeping only what is in default definitions
- add default definitions not in filtered saved dictionary

```
#custom func to check if dictionaries are equal...  from internet
static func are_dictionary_equal(a: Dictionary, b: Dictionary):
	if not a or not b:
		return false
	if a.size() != b.size():
		return false
	for key in a.keys():
		if not b.has(key):
			return false
		if a[key] != b[key]:
			return false
	return true

# return the definition info from the definiiotn.json in the save folder "save name"
static func get_saved_definitions(save_name: String = '') -> Dictionary:
	if save_name == '':
		return load_json(get_config_files_paths()['DEFINITIONS_DEFAULT_SAVE'], get_default_definitions())
		print(" save_name =='' ")
	
	if not save_name in get_saves_folders():
		print("[D] Wasn't able to find save '"+save_name+"'. Loaded the default definitions.")
		return get_default_definitions()
	
	var default_definitions : Dictionary = get_default_definitions()
	
	var saved_definitions : Dictionary = {}
	
	if save_name == '':
		saved_definitions = load_json(get_config_files_paths()['DEFINITIONS_DEFAULT_SAVE'], default_definitions)
		print("definitions data in resource loaded")
		
	else:
		saved_definitions = load_json(WORKING_DIR+"/"+save_name+"/definitions.json", {})
		print("definitions data on ''user://'' loaded") #gamefiend "this one is getting loaded each time I reload a saved game"
	
	# Store variables in arrays from saved and default data
	var base_def_var = default_definitions.variables
	var saved_user_def_var = saved_definitions.variables
	
	# Sort default definitions ids from dialogic editor...
	var sorted_base_id: Array = []
	for key in base_def_var:
		sorted_base_id.append(key.id)
	sorted_base_id.sort()
	
	# Sort saved player definitons ids from user save...
	var sorted_saved_id: Array = []
	for key in saved_user_def_var:
		sorted_saved_id.append(key.id)
	sorted_saved_id.sort()
	
	# Filter saved ids numbers with base id numbers, filter adds only dictionary what is in base def.
	var filtered_saved_ids: Array = []
	for id_sd_num in sorted_saved_id:
		for id_bd_num in sorted_base_id:
			if id_sd_num == id_bd_num:
				filtered_saved_ids.append(id_sd_num)
	
	# Sorting filter_saved_ids by extracting the sorting just the numbers. 
	#I dunno if i need this
	var filtered_saved_ids_num: Array = []
	for key in filtered_saved_ids:
		filtered_saved_ids_num.append(key)
		#print(filtered_saved_ids_num)
	filtered_saved_ids_num.sort()
	
	# Add back into filtered_saves ids string with what is new from base definitions ids strings on disk from editor...
	for missing_data in sorted_base_id:
		var temp_data: String
		for id_sd_num in filtered_saved_ids_num:
			if missing_data == id_sd_num:
				temp_data = id_sd_num
		if temp_data == missing_data:
			pass
		else:
			filtered_saved_ids.append(missing_data)
	
	# Dictionary filter to go through and match saved id string numbers and add only what is in dictionary. 
	#leaves out definitions not in base definitions
	var filtered_saved_definitions: Array = []
	for keys_of_saved_definitions in saved_user_def_var:
		var key_id_of_saved_user_def_var = keys_of_saved_definitions.id
		for id_numbers in filtered_saved_ids:
			if key_id_of_saved_user_def_var == id_numbers:
				#print(key_id_of_saved_user_def_var, " == ", id_numbers)
				filtered_saved_definitions.append(keys_of_saved_definitions)
				
	# Dictionary filter to add new definitons within base definitons
	for temp_base_definitions in base_def_var:
		var ids_base_def:String
		ids_base_def = temp_base_definitions.id
		var saved_ids: String
		var left_over_data
		for keys in filtered_saved_ids_num:
			if ids_base_def == keys:
				saved_ids = keys
		if ids_base_def == saved_ids:
			pass
		else:
			filtered_saved_definitions.append(temp_base_definitions)
			
	# Merge definitions
	var meta2_saved_definitions = saved_definitions.values()
	var meta2_default_definitions = default_definitions.values()
			
	meta2_saved_definitions.sort()
	meta2_default_definitions.sort()

	var arrBox = []
	
	var check = are_dictionary_equal(saved_definitions, default_definitions)
	
	if check == false:
		
		var meta_saved_definitions = saved_definitions.values()
		var meta_default_definitions = default_definitions.values()
		
		
		for i in meta_default_definitions:
			if not i in meta_saved_definitions:
				meta_saved_definitions[0] = meta_default_definitions[0]
				
	return saved_definitions

        #left overs below?...
	#return load_json(WORKING_DIR+"/"+save_name+"/definitions.json", {})
	
```

**Comment by metanoder** [comment]
Edited my comment.  

forgot to put after my filtering and comment out the merging:  
 
saved_definitions.variables = filtered_saved_definitions

**Comment by zaknafean** [comment]
Didn't manage to get to this today, so I'll pick through what you did asap and see what we think!

**Comment by metanoder** [comment]
Hi Zaknafan,

Firstly, **thank you for taking the time tackling updating variables for saved games**.

At the start of this, I didn't know much of anything, nor the built-in functions.  
I think I'd still be parsing through the code for who knows how long without both suggestions, especially Jowan's.  
I'd love it if there's a built in method for this.
   
But after running through this a couple of times, the problem I keep running into is that the dictionaries are nested two levels down. A set of dictionaries in an array, nested in a dictionary.  As you already know, like this...

`{Glossary[], Variables[ {"id" : "123" , "name" : "abc" , "etc" : "etc"  } , {"id" : "456" , "name" : "def" , "etc" : "etc"  } , ... ]`

So say we skip the filtering out old variables loop, 

`saved_definitions.merge(default_definitions, false)` 

It seems to run only on the top leveI and won't go through the nested array data. 
So it will always be equal, {glossary, variables} == {glossary, variables}, and doesn't over write.  
So remains un-updated without new variables from default_definitions pushed out by dialogic editor / developer.

I just ran it again, and the above statement didn't update the 2-level down, nested dictionary sets for me. 

Since it is a nested in an array on nest level two, I have found that using the for loop necessary.  
Unless there is of course a built-in function in array methods I currently don't know.  (admittedly, still a noob)

**On use case for removing old data...**

The use case I was envisioning would be to Retcon a portion of the story, and removing the excess variables no longer in use.  

I am only publishing chapter 1 in my VN, with additional chapters added in the future.  It would give vn creators to retcon their stories.

Necessarily?  Probably not.  Your right, orphaned variables won't do anything harmful.  
But I guess it'd be like having depreciated print statements in the old saves.  No harm, but tidy?  Not so much.

**Comment by metanoder** [comment]
@thebardsrc   

Thanks for taking the time!  
That sounds **phenomenal**!  

If you would humor me, so you end up with one flat dictionary like... 
{ "Top_Level_path" : "Glossary", "nested_glossary_Path" : "nested glossary dictionary path: "data"" , #glossary_etc_etc.., "Top_Level_path" : "Variable" , "nested_Var_Path" : " id: 123" , #variables_etc_etc... }

...I imagine your probably gonna structure it so you don't need the top level dictionary listed...  just paths to each individual dictionary currently nested two levels down...

- This would mean the Path would get around all those nested dictionaries having similar keys...
- able to use .merge()...
- no more need for loops with nested loops.

SOUNDS GREAT!

Refactoring the overall structure and helping to implement this will be the deep end of the pool for me. 
But I'll try to contribute however I can.  

Have a great trip!

Cheers!

**Comment by metanoder** [comment]
Dialogic 2.0 question under similar subject...  

Is this already a feature for Dialogic 2.0?  

Adding variables to an already published game with dialogic saves.  So when users who are already playing my game update, their saved game will reflect added definitions when additional content is being published?

Currently planning to still using 1.x for html publishing.  But once they get HTML export running in Godot 4, the next game I would like to use Dialogic 2 when it's ready

**Comment by thebardsrc** [comment]
> If you would humor me, so you end up with one flat dictionary like...

however I do the very top level im not sure yet, whether its just one biggest one and then the Timelines/Characters/Themes/Varibales/Glossaries are each on that, but what each one would look like from there is like that, yeah. for example with an exceprt from one of my projects, an RPG:

![image](https://user-images.githubusercontent.com/2214793/200676206-c2d9deaa-65ae-4356-a7d8-2936b6a7418e.png)
![image](https://user-images.githubusercontent.com/2214793/200676342-3ede9467-ae99-4d5d-990f-6485321a4947.png)

the dictionary for Variables would then be like this:
'/01-Dahebre/CarterHouse/AutoEvent/Stage' = {id: 923878ridkwhhatever, value: 0}
'/01-Dahebre/CarterHouse/Bedroom/Stage' = not gonna write out the rest from here
'/01-Dahebre/CarterHouse/Jeff/Stage' =
'/01-Dahebre/CarterHouse/Liam/Stage' =
'/01-Dahebre/CarterHouse/Upstairs/Stage' =
'/01-Dahebre/Town/HelicopterChecked' =
'/01-Dahebre/Town/SwordCombo' =
'/Shared/Flags/Test'
'/Test/test2_variety'
'/TrueMyth'
'/Temp'
'/ChangeName'

Tho since you can address variables two ways, with different behaviors each time, I'm gonna ave to check to see what order things are being put in. becuse if I just put [Stage] in there itll pick one, but there's five different paths with that name right now. so it may not be sorted the same way as my example, Ill have to figure out which way it actually puts it currently for compatibity

**Comment by thebardsrc** [comment]
> Dialogic 2.0 question under similar subject...
> 
> Is this already a feature for Dialogic 2.0?
> 
> Adding variables to an already published game with dialogic saves. So when users who are already playing my game update, their saved game will reflect added definitions when additional content is being published?
> 
> Currently planning to still using 1.x for html publishing. But once they get HTML export running in Godot 4, the next game, I would like to use Dialogic 2 when it's ready

Dialogic 2 is a rather different beast, and I don't think saves are even working right now at all 😂 But some of the stuff is  flat paths that way, but Variables are nested dictionaries still. but they dont work the same way, and theyre needed to be that way because you can address them simply in code now. so instead of the function call to access one in Dialogic 1x. you'd simply write it as `var thisvar = Dialogic.VAR.01-Dahebre.CarterHouse.Jeff.Stage'. 

(The 1.x Converter should handle updating all the references in timelines to this automatically, its just other code you gotta change how its done)

But I went and checked how it is right now. Dialogic 2 has all the save data in a somewhat easier to use then 1.x Dictionary called `Dialogic.current_state_info`. when saving, it writes that to a file, when loading it loads that back into a Dictionary and replaces the previous current_state_info in memory, then starts to load back each subsystem's state

For variables specifically, in 2.0 those are being stored in the Godot.project file, instead of a JSON like in 1.x. It's written like this: 

```
func load_game_state():
	dialogic.current_state_info['variables'] = merge_folder(dialogic.current_state_info['variables'], DialogicUtil.get_project_setting('dialogic/variables', {}))


func merge_folder(new, defs) -> Dictionary:
	# also go through all groups in this folder
	for x in new.keys():
		if x in defs and typeof(new[x]) == TYPE_DICTIONARY:
			new[x] = merge_folder(new[x], defs[x])
	# add all new variables
	for x in defs.keys():
		if not x in new:
			new[x] = defs[x]
	return new

```

Unless I'm reading that wrong just looking at it, it starts with the save file's variable state, and then if there's a variable name in the Project settigns that's not already in there then it adds it to the dictionary

**Comment by metanoder** [comment]
Just in case the refactor and the 1.x speed ups are coming later rather than sooner, I initiated a pull request for the get_saved_defintions I modified.  A stop gap for those gamers who wanna push out updates / extend their vn games. #1334

**Comment by metanoder** [comment]
@zaknafean Thanks for reviewing and merging pr #1334 .  This closes this issue.

**Comment by thebardsrc** [comment]
reoepning just so i can verify and check back off that the changes in #1271  actually work properly for this as I originally intended

**Comment by thebardsrc** [comment]
confirmed working proplery in #1271 , so closing again
